lib/model/typing_status.dart

```diff
@@ -0,0 +1,83 @@
+import 'dart:async';
+
+import 'package:flutter/foundation.dart';
+
+import '../api/model/events.dart';
+import 'narrow.dart';
+
+/// The view-model for tracking the typing status organized by narrows.
+class TypingStatus extends ChangeNotifier {
+  @visibleForTesting
+  final Map<SendableNarrow, Map<int, Timer>> typistIdsByNarrow = {};
+
+  final int selfUserId;
+  final Duration typingStartedExpiryPeriod;
+  final Duration typingStoppedWaitPeriod;
+  final Duration typingStartedWaitPeriod;
+
+  TypingStatus({
+    required this.selfUserId,
+    required this.typingStartedExpiryPeriod,
+    required this.typingStoppedWaitPeriod,
+    required this.typingStartedWaitPeriod,
+  });
+
+  @override
+  void dispose() {
+    for (final typistTimer in typistIdsByNarrow.values) {
```

Comment:
`typistTimer` doesn't quite fit as a name for a `Map<int, Timer>`. Maybe `timersForNarrow`?